File: TestMatrix.java
import java.util.regex.*;

public class TestMatrix {
    public static void main(String[] args) {
        String s = "\\begin{pmatrix}\n a & b \\\\\n c & d\n\\end{pmatrix}";
        
        s = s.replace("\\left", "").replace("\\right", "");
        s = s.replace("\\{", "{").replace("\\}", "}");
        s = s.replace("\\,", " ").replace("\\!", "").replace("\\;", " ").replace("\\:", " ");
        s = s.replace("\\\\", "\n");
        
        Pattern p = Pattern.compile("\\\\begin\\{(p?matrix|b?matrix|Bmatrix|vmatrix|Vmatrix|array)\\}([\\s\\S]*?)\\\\end\\{\\1\\}", Pattern.DOTALL);
        Matcher m = p.matcher(s);
        StringBuffer sb = new StringBuffer();
        while(m.find()) {
            String inner = m.group(2);
            String[] rows = inner.split("\n"); // in Kotlin I used \\\\ which splits by \\, but wait! In Kotlin I used "\\\\" to split!
            StringBuilder table = new StringBuilder("[\n");
            for(String row : rows) {
                if(row.trim().isEmpty()) continue;
                String[] cols = row.split("&");
                table.append(String.join(" | ", cols)).append("\n");
            }
            table.append("]");
            m.appendReplacement(sb, Matcher.quoteReplacement(table.toString()));
        }
        m.appendTail(sb);
        s = sb.toString();
        
        System.out.println(s);
    }
}

File: test_matrix.kt
fun main() {
    var s = "\\begin{pmatrix}\n a & b \\\\\n c & d\n\\end{pmatrix}"
    
    // Step 1
    s = s.replace("\\left", "").replace("\\right", "")
    s = s.replace("\\{", "{").replace("\\}", "}")
    s = s.replace("\\,", " ").replace("\\!", "").replace("\\;", " ").replace("\\:", " ")
    s = s.replace("\\\\", "\n") // LaTeX line break

    // Step 1b
    val matrixEnvRegex = Regex("\\\\begin\\{(p?matrix|b?matrix|Bmatrix|vmatrix|Vmatrix|array)\\}([\\s\\S]*?)\\\\end\\{\\1\\}", RegexOption.DOT_MATCHES_ALL)
    s = matrixEnvRegex.replace(s) { m ->
        val rows = m.groupValues[2].split("\\\\").map { row ->
            row.split("&").joinToString(" | ") { it.trim() }
        }.filter { it.isNotBlank() }
        "[\n" + rows.joinToString("\n") + "\n]"
    }
    
    println(s)
}
